src/plugins/action-queue/types.js

```diff
@@ -0,0 +1,171 @@
+
+//
+// Action Operations
+//
+
+export type ActionOp =
+  | {
+      type: 'seq',
+      actions: ActionOp[]
+    }
+  | {
+      type: 'par',
+      actions: ActionOp[]
+    }
+  | {
+      type: 'exchange:buy',
+      nativeAmount: string,
+      walletId: string,
+      tokenId?: string,
+      exchangePluginId: string
+    }
+  | {
+      type: 'exchange:sell',
+      nativeAmount: string,
+      walletId: string,
+      tokenId?: string,
+      exchangePluginId: string
+    }
+  | {
+      type: 'loan:borrow',
+      borrowPluginId: string,
+      nativeAmount: string,
+      walletId: string,
+      tokenId?: string
+    }
+  | {
+      type: 'loan:deposit',
+      borrowPluginId: string,
+      nativeAmount: string,
+      walletId: string,
+      tokenId?: string
+    }
+  | {
+      type: 'loan:repay',
+      borrowPluginId: string,
+      nativeAmount: string,
+      walletId: string,
+      tokenId?: string
+    }
+  | {
+      type: 'loan:withdraw',
+      borrowPluginId: string,
+      nativeAmount: string,
+      walletId: string,
+      tokenId?: string
+    }
+  | {
+      type: 'swap',
+      fromWalletId: string,
+      toWalletId: string,
+      fromTokenId?: string,
+      toTokenId?: string,
+      nativeAmount: string
+    }
+  // Useful for development/testing
+  | {
+      type: 'toast',
+      message: string
+    }
+  | {
+      type: 'delay',
+      ms: number
+    }
+
+//
+// Action Effects
+//
+
+export type ActionEffect =
+  | {
+      type: 'seq',
+      opIndex: number,
+      childEffect: ActionEffect
+    }
+  | {
+      type: 'par',
+      childEffects: ActionEffect[]
+    }
+  | {
+      type: 'balance',
+      address?: string, // Address balance if present otherwise wallet balances
+      aboveAmount?: string,
+      belowAmount?: string,
+      walletId: string,
+      tokenId?: string
+    }
+  | {
+      type: 'tx-confs',
+      txId: string,
+      walletId: string,
+      confirmations: number
+    }
+  | {
+      type: 'price',
+      currencyPair: string,
+      aboveRate?: string,
+      belowRate?: string
+    }
+  | {
+      type: 'end',
+      error?: Error
+    }
+  // Useful for development/testing
+  | {
+      type: 'unixtime',
+      timestamp: number
+    }
+
+//
+// Action Program
+//
+
+// On Disk:
+export type ActionProgram = {
+  programId: string,
+  actionOp: ActionOp
+}
+
+export type ActionProgramState = {
+  programId: string,
+  effect?: ActionEffect
+}
+
+//
+// Internal Types
+//
+
+export type ActionQueueItem = {
+  program: ActionProgram,
+  state: ActionProgramState
+}
+export type ActionQueueMap = {
+  [id: string]: ActionQueueItem
+}
+
+//
+// Action API
+//
+
+declare export function makeActionProgram(actionOp: ActionOp): Promise<ActionProgram>
+
+declare function executeActionOp(actionOp: ActionOp): ActionEffect
+
+//
+// Aciton Display API
+//
+
+export type ActionDisplayInfo = {
+  title: string,
+  message: string,
+  complete: boolean,
```

Comment:
Doesn't the Figma design support multiple states? not-started / started / done / error

Reply:
Needs to be changed to `status: 'pending' | 'complete' | Error` (something I am going to do today).

The design doesn't include a "incomplete" state, and it doesn't even include designs for error states (yet). Do you think we need to design for an "incomplete" vs "pending" state, where the former means we haven't done the action and the latter is "we have done the action but we're waiting on the the effect"? The "incomplete" state is immediately after the previous action's effect is observed, so the time between "effect observed" and "do next action" is milliseconds, so I wouldn't think this state should be tracked from the GUI. If this state is omitted, "pending" would stand in for both states (an action is "pending" if the program is blocked by its effect or a previous effect before it).

Reply:
`pending | doing | done | Error`